File: sdks/node/src/resources/webhooks.ts
/**
 * HookSniff SDK — Webhooks Resource (Send webhooks)
 */

import { HttpMethod, HookSniffRequest, type HookSniffRequestContext } from "../request";
import type { PostOptions } from "../util";
import type {
  CreateWebhookRequest,
  BatchWebhookRequest,
  Delivery,
  DeliveryListResponse,
  BatchResponse,
  BatchReplayRequest,
} from "../models";
import { paginatedIterator, type PaginationOptions } from "../pagination";

export interface WebhookSendOptions extends PostOptions {}

export interface WebhookListOptions extends PaginationOptions {
  endpoint_id?: string;
  status?: string;
}

export class Webhooks {
  constructor(private readonly requestCtx: HookSniffRequestContext) {}

  /** Send a single webhook. */
  public send(body: CreateWebhookRequest, options?: WebhookSendOptions): Promise<Delivery> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/webhooks");
    request.setBody(body);
    if (options?.idempotencyKey) {
      request.setHeaderParam("idempotency-key", options.idempotencyKey);
    }
    return request.send(this.requestCtx, (json) => json as Delivery);
  }

  /** Send multiple webhooks in a batch. */
  public sendBatch(body: BatchWebhookRequest, options?: PostOptions): Promise<BatchResponse> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/webhooks/batch");
    request.setBody(body);
    if (options?.idempotencyKey) {
      request.setHeaderParam("idempotency-key", options.idempotencyKey);
    }
    return request.send(this.requestCtx, (json) => json as BatchResponse);
  }

  /** Get a delivery by ID. */
  public getDelivery(deliveryId: string): Promise<Delivery> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/webhooks/{id}");
    request.setPathParam("id", deliveryId);
    return request.send(this.requestCtx, (json) => json as Delivery);
  }

  /** List deliveries. */
  public listDeliveries(options?: WebhookListOptions): Promise<DeliveryListResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/webhooks");
    request.setQueryParams({
      limit: options?.limit,
      iterator: options?.iterator,
      endpoint_id: options?.endpoint_id,
      status: options?.status,
    });
    return request.send(this.requestCtx, (json) => json as DeliveryListResponse);
  }

  /** Auto-paginate through all deliveries. */
  public listAllDeliveries(options?: WebhookListOptions): AsyncIterable<Delivery> {
    return paginatedIterator(
      (opts) => this.listDeliveries({ ...options, ...opts }),
      options
    );
  }

  /** Replay a specific delivery. */
  public replay(deliveryId: string): Promise<Delivery> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/webhooks/{id}/replay");
    request.setPathParam("id", deliveryId);
    return request.send(this.requestCtx, (json) => json as Delivery);
  }

  /** Batch replay multiple deliveries. */
  public replayBatch(body: BatchReplayRequest): Promise<BatchResponse> {
    const request = new HookSniffRequest(HttpMethod.POST, "/v1/webhooks/batch/replay");
    request.setBody(body);
    return request.send(this.requestCtx, (json) => json as BatchResponse);
  }

  /** Get delivery attempts for a delivery. */
  public getAttempts(deliveryId: string): Promise<import("../models").DeliveryAttempt[]> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/webhooks/{id}/attempts");
    request.setPathParam("id", deliveryId);
    return request.send(this.requestCtx, (json) => json as import("../models").DeliveryAttempt[]);
  }
}

File: sdks/java/src/main/java/com/hooksniff/resources/Webhooks.java
package com.hooksniff.resources;

import com.hooksniff.ApiException;
import com.hooksniff.HookSniffConfig;
import com.hooksniff.RequestHelper;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

/**
 * Send, list, get, replay, and batch webhooks.
 */
public class Webhooks {
    private static final Type DELIVERY = new TypeToken<Map<String, Object>>(){}.getType();
    private static final Type DELIVERY_LIST = new TypeToken<Map<String, Object>>(){}.getType();
    private static final Type BATCH = new TypeToken<Map<String, Object>>(){}.getType();

    private final HookSniffConfig config;

    public Webhooks(HookSniffConfig config) {
        this.config = config;
    }

    /** Send a single webhook */
    public Map<String, Object> send(Map<String, Object> input) throws ApiException, IOException, InterruptedException {
        return send(input, null);
    }

    /** Send a single webhook with idempotency key */
    public Map<String, Object> send(Map<String, Object> input, String idempotencyKey) throws ApiException, IOException, InterruptedException {
        RequestHelper req = new RequestHelper("POST", "/v1/webhooks");
        if (idempotencyKey != null) req.setHeaderParam("idempotency-key", idempotencyKey);
        req.setBody(input);
        return req.send(config, DELIVERY);
    }

    /** Send batch webhooks */
    public Map<String, Object> batch(Map<String, Object> input) throws ApiException, IOException, InterruptedException {
        return batch(input, null);
    }

    /** Send batch webhooks with idempotency key */
    public Map<String, Object> batch(Map<String, Object> input, String idempotencyKey) throws ApiException, IOException, InterruptedException {
        RequestHelper req = new RequestHelper("POST", "/v1/webhooks/batch");
        if (idempotencyKey != null) req.setHeaderParam("idempotency-key", idempotencyKey);
        req.setBody(input);
        return req.send(config, BATCH);
    }

    /** List deliveries */
    public Map<String, Object> list() throws ApiException, IOException, InterruptedException {
        return list(null, null);
    }

    /** List deliveries with pagination */
    public Map<String, Object> list(Integer limit, Integer offset) throws ApiException, IOException, InterruptedException {
        RequestHelper req = new RequestHelper("GET", "/v1/webhooks");
        java.util.HashMap<String, Object> params = new java.util.HashMap<>();
        if (limit != null) params.put("limit", limit);
        if (offset != null) params.put("offset", offset);
        if (!params.isEmpty()) req.setQueryParams(params);
        return req.send(config, DELIVERY_LIST);
    }

    /** Get a specific delivery */
    public Map<String, Object> get(String id) throws ApiException, IOException, InterruptedException {
        RequestHelper req = new RequestHelper("GET", "/v1/webhooks/{id}");
        req.setPathParam("id", id);
        return req.send(config, DELIVERY);
    }

    /** Replay a delivery */
    public Map<String, Object> replay(String id) throws ApiException, IOException, InterruptedException {
        return replay(id, null);
    }

    /** Replay a delivery with idempotency key */
    public Map<String, Object> replay(String id, String idempotencyKey) throws ApiException, IOException, InterruptedException {
        RequestHelper req = new RequestHelper("POST", "/v1/webhooks/{id}/replay");
        req.setPathParam("id", id);
        if (idempotencyKey != null) req.setHeaderParam("idempotency-key", idempotencyKey);
        return req.send(config, DELIVERY);
    }
}
